packages/ckb-indexer/src/ckbIndexerFilter.ts

```diff
@@ -0,0 +1,352 @@
+          blockchain.Script.pack(cellOutput.lock),
+          blockchain.Script.pack(script)
+        )
+      ) {
+        return false;
+      }
+    } else {
+      if (
+        !cellOutput.type ||
+        !bytes.equal(
+          blockchain.Script.pack(cellOutput.type),
+          blockchain.Script.pack(script)
+        )
+      ) {
+        return false;
+      }
+    }
+    // Prefix mode
+  } else {
+    if (scriptType === "lock") {
+      if (!checkScriptWithPrefixMode(cellOutput.lock, script)) {
+        return false;
+      }
+    } else {
+      if (!checkScriptWithPrefixMode(cellOutput.type, script)) {
+        return false;
+      }
+    }
+  }
+
+  // the "exact" mode works only on "SearchKey.script",
+  // not on "SearchKey.filter.script"
+  // the "SearchKey.filter.script" is always in prefix mode
+  if (!filter) return true;
+  // filter type script if scriptType is "lock"
+  if (scriptType === "lock") {
+    if (
+      filter.script &&
+      !checkScriptWithPrefixMode(cellOutput.type, filter.script)
+    ) {
+      return false;
+    }
+
+    if (
+      filter.scriptLenRange &&
+      !checkScriptLenRange(cellOutput.type, filter.scriptLenRange)
+    ) {
+      return false;
+    }
+    // filter lock script if scriptType is "type"
+  } else {
+    if (
+      filter.script &&
+      !checkScriptWithPrefixMode(cellOutput.lock, filter.script)
+    ) {
+      return false;
+    }
+    if (
+      filter.scriptLenRange &&
+      !checkScriptLenRange(cellOutput.lock, filter.scriptLenRange)
+    ) {
+      return false;
+    }
+  }
+
+  const { blockRange, outputCapacityRange, outputDataLenRange } = filter;
+
+  if (blockRange) {
+    const blockNumber = BI.from(cell.blockNumber || Number.MAX_SAFE_INTEGER);
+    const fromBlock = BI.from(blockRange[0]);
+    const toBlock = BI.from(blockRange[1]);
+    if (blockNumber.lt(fromBlock) || blockNumber.gte(toBlock)) {
+      return false;
+    }
+  }
+
+  if (outputCapacityRange) {
+    const capacity = BI.from(cellOutput.capacity);
+    const fromCapacity = BI.from(outputCapacityRange[0]);
+    const toCapacity = BI.from(outputCapacityRange[1]);
+    if (capacity.lt(fromCapacity) || capacity.gte(toCapacity)) {
+      return false;
+    }
+  }
+
+  if (outputDataLenRange) {
+    const dataLen = BI.from(bytes.bytify(cell.data).length);
+    const fromDataLen = BI.from(outputDataLenRange[0]);
+    const toDataLen = BI.from(outputDataLenRange[1]);
+    if (dataLen.lt(fromDataLen) || dataLen.gte(toDataLen)) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+function checkScriptWithPrefixMode(
+  script: Script | undefined,
+  filterScript: Script
+): boolean {
+  if (!script) {
+    return false;
+  }
+
+  const expectCodeHashPrefix = bytes.bytify(filterScript.codeHash);
+  const actualCodeHashPrefix = bytes
+    .bytify(script.codeHash)
+    .slice(0, expectCodeHashPrefix.length);
+  if (!bytes.equal(expectCodeHashPrefix, actualCodeHashPrefix)) {
+    return false;
+  }
```

Comment:
the `codeHash` can only be exact match, since the `codeHash` is always 32 bytes

Reply:
updated